single_step_avd/src/merkle_tree_avd/mod.rs

```diff
@@ -1,45 +1,53 @@
 use rand::Rng;
-use sha3::{digest::Digest, Sha3_256};
-
-use std::{collections::HashMap, error::Error as ErrorTrait, fmt};
-
-use crate::{Error, SingleStepAVD};
+use sha3::{
+    digest::Digest,
+    Sha3_256
+};
+use std::{
+    error::Error as ErrorTrait,
+    fmt
+};
+use crate::{
+    Error,
+    SingleStepAVD
+};
 use crypto_primitives::{
     sparse_merkle_tree::{
-        MerkleDepth, MerkleIndex, MerkleTreeParameters, MerkleTreePath, SparseMerkleTree,
+        store::SMTStorer,
+        MerkleDepth,
+        MerkleIndex,
+        MerkleTreeParameters,
+        MerkleTreePath,
     },
     hash::FixedLengthCRH,
 };
 
+pub mod store;
 pub mod constraints;
 
 pub trait MerkleTreeAVDParameters {
     const MAX_UPDATE_BATCH_SIZE: u64;
     const MAX_OPEN_ADDRESSING_PROBES: u8;
-
-    type MerkleTreeParameters: MerkleTreeParameters;
+    type SMTStorer: SMTStorer;
 
     fn is_valid(&self) -> Result<bool, Error> {
-        Self::MerkleTreeParameters::is_valid()
+        <<Self as MerkleTreeAVDParameters>::SMTStorer as SMTStorer>::P::is_valid()
     }
 }
 
-pub struct MerkleTreeAVD<P: MerkleTreeAVDParameters> {
-    tree: SparseMerkleTree<P::MerkleTreeParameters>,
-    key_d: HashMap<[u8; 32], (u8, u64, [u8; 32])>,
-    // key -> probe, version, value
-    index_d: HashMap<MerkleIndex, [u8; 32]>,
+pub struct MerkleTreeAVD<T: store::MTAVDStorer> {
```

Comment:
@nirvantyagi This compiler warning looks similar to other:

he function or associated item `new` exists for struct `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::constraints::tests::MerkleTreeAVDTestParameters>`, but its trait bounds were not satisfied function or associated item `new` not found for this note: the following trait bounds were not satisfied: `merkle_tree_avd::constraints::tests::MerkleTreeAVDTestParameters: merkle_tree_avd::store::MTAVDStorer` which is required by `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::constraints::tests::MerkleTreeAVDTestParameters>: SingleStepAVD` help: items from traits can only be used if the trait is implemented and in scope
the function or associated item `new` exists for struct `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::constraints::tests::MerkleTreeAVDTestParameters>`, but its trait bounds were not satisfied doesn't satisfy `_: SingleStepAVD` note: the following trait bounds were not satisfied: `merkle_tree_avd::constraints::tests::MerkleTreeAVDTestParameters: merkle_tree_avd::store::MTAVDStorer` which is required by `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::constraints::tests::MerkleTreeAVDTestParameters>: SingleStepAVD` help: items from traits can only be used if the trait is implemented and in scope
the function or associated item `new` exists for struct `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>`, but its trait bounds were not satisfied function or associated item `new` not found for this note: the following trait bounds were not satisfied: `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::digest::Update` which is required by `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::digest::FixedOutput` which is required by `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::digest::Reset` which is required by `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: std::clone::Clone` which is required by `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: std::default::Default` which is required by `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `merkle_tree_avd::tests::MerkleTreeAVDTestParameters: merkle_tree_avd::store::MTAVDStorer` which is required by `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: SingleStepAVD` `&merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::digest::Update` which is required by `&merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `&merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::digest::FixedOutput` which is required by `&merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `&merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::digest::Reset` which is required by `&merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `&merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: std::default::Default` which is required by `&merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `&mut merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::digest::Update` which is required by `&mut merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `&mut merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::digest::FixedOutput` which is required by `&mut merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `&mut merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::digest::Reset` which is required by `&mut merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `&mut merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: std::clone::Clone` which is required by `&mut merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` `&mut merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: std::default::Default` which is required by `&mut merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>: sha3::Digest` help: consider annotating `merkle_tree_avd::MerkleTreeAVD<merkle_tree_avd::tests::MerkleTreeAVDTestParameters>` with `#[derive(std::clone::Clone, std::default::Default)]`: `#[derive(std::clone::Clone, std::default::Default)] `